### DISCUSSION TITLE
Integrate mosparo into keycloak

### DISCUSSION THREAD

**Discussion by xundeenergie** [mosparo/mosparo#187] [Integrations]
I did not find any information, how to integrate mosparo into keycloak as alternative to googles recapchta.
I found one for hcaptcha https://github.com/p08dev/keycloak-hcaptcha but i have no idea, how to do this for mosparo.

**Comment by zepich** [comment]
Hi @xundeenergie 

Thank you very much for your question.

The hCaptcha provider is a perfect example how to do it.

In general, it must be possible to configure the connection to a mosparo installation (like in the hCaptcha configuration screen https://github.com/p08dev/keycloak-hcaptcha/blob/master/img/step-03.png).

After that, the mosparo frontend code must be added to the template. I guess that's done here: https://github.com/p08dev/keycloak-hcaptcha/blob/master/img/step-05.png

And then, the code must connect the settings with the template (providing all the required information from the config page to the template). 

Of course, the code must verify the data with the mosparo installation. That's done in the code here, I guess: https://github.com/p08dev/keycloak-hcaptcha/blob/master/src/main/java/de/itrupp/p8/keycloak/authenticator/RegistrationhCaptcha.java#L156

If you're a developer, we can work through the creation of an integration for Keycloak. If you're not a developer, that's fine. I will have a closer look at it in the next few days.

Kind regards,

zepich

**Reply by xundeenergie** [reply]
Hi!
That sounds great!

I orientated on this howto for hcaptcha... but i think, mosparo works a little other... hcaptcha is integrated with a iframe and mosparo works with js directly integrated in the website. How i understood the docu... 

What i've understood, there must be the div with id mosparo-box and a javascript which replaces this div with mosparos frontend.

The verification, how i understood mosparos docu, has to be done in several steps what has to be done by the keycloak spi, which is handled by hcaptcha i the iframe. 

Or is it possible to load an iframe for mosparo?

**Reply by zepich** [reply]
Hi @xundeenergie 

In general, the implementation should be more or less the same as with hCaptcha.

1. You have to include the frontend part. For hCaptcha it's a `<div>` (see here: https://github.com/p08dev/keycloak-hcaptcha, step "Registration template") and the provider will add the JavaScript file from hCaptcha automatically (https://github.com/p08dev/keycloak-hcaptcha/blob/master/src/main/java/de/itrupp/p8/keycloak/authenticator/RegistrationhCaptcha.java#L122). For mosparo you have to add an empty `<div>` comparable to the one of hCaptcha and you have to add an external JavaScript script (`/build/mosparo-frontend.js`). Additionally, you have to add a local script to initialize the empty div with the mosparo settings (that's all visible here: https://documentation.mosparo.io/docs/integration/custom#add-container and https://documentation.mosparo.io/docs/integration/custom#embedding-javascript).
2. The user has to click the checkbox so that mosparo can do the frontend validation (equal to hCaptcha, where the user has to click the checkbox too so hCaptcha can verify the user on their server).
3. The backend (in your case, Keycloak) has to verify that the user has clicked the checkbox (same for hCaptcha and mosparo). Here, the process is a bit more complicated than hCaptcha since mosparo needs the submitted form data as SHA256 hash and with some signatures to make sure everything works perfectly. This process has to be started in the Keycloak provider, like here: https://github.com/p08dev/keycloak-hcaptcha/blob/master/src/main/java/de/itrupp/p8/keycloak/authenticator/RegistrationhCaptcha.java#L156). This process is the more complicated part since all the hashes and signatures must be generated. The documentation describes this process step by step here: https://documentation.mosparo.io/docs/integration/custom#adding-mosparo-to-the-process). The example code is written in PHP, but adopting that code for Java should be no problem. The eleven steps will go through everything you must do to verify the form data with mosparo. The verification part is a bit more complicated than hCaptcha, but it's required since we do not use `magic` to verify the form - we use the form data to ensure there is no spam.

Please let me know if you have any questions.

Kind regards,

zepich

**Reply by xundeenergie** [reply]
Ok.
I did the same modification in the template like in hcaptcha. I also give the div the id mosparo-box... but the captcha is not loaded... 
What's the best way, to show you my approach?

I can create a repo here for my provider... or you can have look on my personal forgejo-git-instance.

**Reply by zepich** [reply]
> I can create a repo here for my provider... or you can have look on my personal forgejo-git-instance.

That depends on you, to be honest. For me, all options are okay. Here, on GitHub, you can create a public or a private repository and add my username as editor, but if you prefer your own git instance, that's also okay for me. You can also send this information privately by email to me (feedback@mosparo.io).

**Reply by xundeenergie** [reply]
Would be nice, if you can help me here
https://github.com/xundeenergie/keycloak-mosparo

or if you send me your public-key, i can add you to my forgejo-git.

**Reply by zepich** [reply]
Public key: 
```
ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIBmKk/vvudrXlZZn02bcpkEhbvk2zTpMmbq/jwWbBAAZ zepi@zepi.net
```

**Reply by xundeenergie** [reply]
Added your key to https://git.schuerz.at/jakob/keycloak-mosparo 
You can now push and pull.

**Reply by zepich** [reply]
Hi @xundeenergie 

Thank you very much for the repo.

I've just pushed the provider with the mosparo verification. You have to build it with Maven as described in the README and enable it in the registration form. It's not 100% finished and perfect code, but it works with Keycloak 23.0.6 (at least on my machine ;) ).

I'm looking forward to your feedback.

Kind regards,

zepich

**Reply by xundeenergie** [reply]
Hi!

Thank you. Was a little bit irritated, because you pushed, while i was working on it... :laughing: 

You use the method "getFirst", which is not resolveable. I replaced it with "get(0)", but i'm not sure if it is, what you are intended.
Its Line 174 and 187, where you used entry.getValue().getFirst();

And there was a small issue with mosparo_host, what i fixed (not pushed now)

When i send the registration now, i get a 400 Bad Request.
In keycloak log i see

```
Feb 18 01:31:00 ldap kc.sh[557663]: 2024-02-18 01:31:00,087 WARN  [org.keycloak.services] (executor-thread-6) KC-SERVICES0013: Failed authentication: java.lang.IndexOutOfBoundsException: Index: 1, Size: 1
Feb 18 01:31:00 ldap kc.sh[557663]:         at java.base/java.util.LinkedList.checkElementIndex(LinkedList.java:559)
Feb 18 01:31:00 ldap kc.sh[557663]:         at java.base/java.util.LinkedList.get(LinkedList.java:480)
Feb 18 01:31:00 ldap kc.sh[557663]:         at at.schuerz.keycloak.authenticator.RegistrationMosparo.verifyFormData(RegistrationMosparo.java:174)
Feb 18 01:31:00 ldap kc.sh[557663]:         at at.schuerz.keycloak.authenticator.RegistrationMosparo.validate(RegistrationMosparo.java:140)
Feb 18 01:31:00 ldap kc.sh[557663]:         at org.keycloak.authentication.FormAuthenticationFlow.processAction(FormAuthenticationFlow.java:215)
Feb 18 01:31:00 ldap kc.sh[557663]:         at org.keycloak.authentication.DefaultAuthenticationFlow.processAction(DefaultAuthenticationFlow.java:133)
Feb 18 01:31:00 ldap kc.sh[557663]:         at org.keycloak.authentication.AuthenticationProcessor.authenticationAction(AuthenticationProcessor.java:988)
Feb 18 01:31:00 ldap kc.sh[557663]:         at org.keycloak.services.resources.LoginActionsService.processFlow(LoginActionsService.java:362)
Feb 18 01:31:00 ldap kc.sh[557663]:         at org.keycloak.services.resources.LoginActionsService.processRegistration(LoginActionsService.java:706)
Feb 18 01:31:00 ldap kc.sh[557663]:         at org.keycloak.services.resources.LoginActionsService.registerRequest(LoginActionsService.java:762)
Feb 18 01:31:00 ldap kc.sh[557663]:         at org.keycloak.services.resources.LoginActionsService.processRegister(LoginActionsService.java:740)
Feb 18 01:31:00 ldap kc.sh[557663]:         at org.keycloak.services.resources.LoginActionsService$quarkusrestinvoker$processRegister_707378c0de357d574c0a2e1d6056afe7606ab5ff.invoke(Unknown Source)
Feb 18 01:31:00 ldap kc.sh[557663]:         at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)
Feb 18 01:31:00 ldap kc.sh[557663]:         at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:141)
Feb 18 01:31:00 ldap kc.sh[557663]:         at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:145)
Feb 18 01:31:00 ldap kc.sh[557663]:         at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:576)
Feb 18 01:31:00 ldap kc.sh[557663]:         at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
Feb 18 01:31:00 ldap kc.sh[557663]:         at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
Feb 18 01:31:00 ldap kc.sh[557663]:         at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
Feb 18 01:31:00 ldap kc.sh[557663]:         at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
Feb 18 01:31:00 ldap kc.sh[557663]:         at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
Feb 18 01:31:00 ldap kc.sh[557663]:         at java.base/java.lang.Thread.run(Thread.java:840)
Feb 18 01:31:00 ldap kc.sh[557663]: 2024-02-18 01:31:00,089 WARN  [org.keycloak.events] (executor-thread-6) type="REGISTER_ERROR", realmId="545fa198-0b4f-4c01-84eb-61eaba9b71df", clientId="events.fedibikes.social", userId="null", ipAddress="XXX.YYY.ZZZ.AAA", error="invalid_user_credentials", auth_method="openid-connect", auth_type="code", register_method="form", last_name="Guckindiluft", redirect_uri="https://my.app.example/auth/keycloak/callback", first_name="Hans", code_id="b2ea85ee-9eeb-4818-baec-e2134dd56d7d", email="my@email.tld", username="shahsie4"
```

**Reply by xundeenergie** [reply]
Oh, i got it... i tried "get(1)"... now i use "get(0)" and get another error on registration:

```
Feb 18 01:56:26 ldap kc.sh[558735]: 2024-02-18 01:56:26,084 WARN  [org.keycloak.events] (executor-thread-6) type="REGISTER_ERROR", realmId="73ed6f65-c8e5-4551-93fb-49ad89d20b21", clientId="app.client.id", userId="null", ipAddress="194.24.159.93", error="invalid_registration", auth_method="openid-connect", register_method="form", last_name="Foo", redirect_uri="https://myapp.example.com/auth/keycloak/callback", first_name="Bertl", code_id="b2ea85ee-9eeb-4818-baec-e2134dd56d7d", email="my@email.tld", username="lalelu"
```

This happens, when i send the registration finally.
The "userId" is "null"... i think, this could be a problem?
The name in the form is "username".

I pushed the partly fixed version to the branch "partly-working", if you want check, what i've done.

**Reply by zepich** [reply]
Hi @xundeenergie 

My committed version from yesterday fully works with Keycloak 23.0.6 - at least on my machine. What version of Keycloak do you use?

My feedback:

1. The error message you sent makes no sense. I think you haven't tested my version, because there is no `getFirst()` in my version in line 174. Can you verify that you tested my version?
2. There is a merge conflict stuck in the committed file. You should resolve that one https://git.schuerz.at/jakob/keycloak-mosparo/src/branch/master/src/main/java/at/schuerz.keycloak/authenticator/RegistrationMosparo.java#L157
3. I don't know exactly why you removed these lines (https://git.schuerz.at/jakob/keycloak-mosparo/commit/33aec2d52c88f930c732fe33de7ad8b26be08f61, Lines 81, 82, and 122), but they are required since we need the private key to verify the form data.
4. In the `partly-working` branch, I can see my first commit from Thursday, but there is no recent commit. Can you verify that you've pushed the latest commits there? 
5. I recommend setting the `https://` in the configuration field and not hardcoded here because then you can copy the values for the configuration 1:1 from your mosparo installation (https://git.schuerz.at/jakob/keycloak-mosparo/commit/d94b801c965b72bb3286e14690b0b631ab137d27).
![image](https://github.com/mosparo/mosparo/assets/391914/3dbcf8c0-638a-449e-b5a3-48b7b7c53af8)

Please clean up the committed version and test it again. Please let me know what line the error message appears after the cleanup.

Kind regards,

zepich

**Reply by xundeenergie** [reply]
It's strange... seems to be a java-problem.
Which java do you use to compile it?

This shows my intellij
![image](https://github.com/mosparo/mosparo/assets/5894829/31530631-7ac0-4821-9af9-a1702e0e53b6)

I use JAVA 17
Keycloak is 23.0.6

**Reply by zepich** [reply]
Okay, that's the problem. I use OpenJDK 21.0.2.

I will test it out with v17.

**Reply by xundeenergie** [reply]
I had to do some small modifications.
Please look at https://git.schuerz.at/jakob/keycloak-mosparo/src/branch/working

Now it compiles and works for me.

**Reply by xundeenergie** [reply]
In debian the newest openjdk is 17 :)

**Comment by xundeenergie** [comment]
Had some rebase-troubles... 
Now i recreated the repo new and have 2 branches. 
Master for JAVA 17
java21 for JAVA 21

I also pushed the repo to github as mirror.

And now i will update the README.

Thank you for your work and help. @zepich

**Reply by zepich** [reply]
Hi @xundeenergie 

Amazing! Thank you very much for your feedback and work. I've tested the `master` branch with JDK 21, and it works perfectly fine. I think you can maintain one branch to make it easier to use it.

Kind regards,

zepich

**Comment by xundeenergie** [comment]
I've a new problem with mosparo.
But i don't know, when it comes into the system.

What i see is, the mosparo javascript "mosparo-frontend.js" loads successfully. But the script is not executed and the fields, for example _mosparo_validationToken is not created. In case of that, the mosparo-validation fails in keycloak.

Maybe a keycloak-update broke this flow.

So... the same... hcaptcha-flow must be broken too?

**Comment by xundeenergie** [comment]
could solve it. Changed it in the readme of my repo.

https://github.com/xundeenergie/keycloak-mosparo
https://git.schuerz.at/jakob/keycloak-mosparo

**Comment by zepich** [comment]
Hi @xundeenergie 

Thank you very much for your post.

I guess the frontend JavaScript initialization code was missing or had a problem, correct? Or what was exactly the problem (just that others know what they have to do if they experience the same issue)?

Kind regards,

zepich

**Comment by xundeenergie** [comment]
Yes. I forgot to modify the theme in the jar-file... 

And i had to add "https://" in front of {mosparoHost}, because it redirected it to suburl under my keycloak. 
Added this to my keycloak-mosparo repo(s) i linked above.

now it works

**Reply by zepich** [reply]
Perfect, thank you very much for the information!
